examples/client.rs

```diff
@@ -1,41 +1,100 @@
+use std::mem;
 
 use audrey::read::Reader;
 use audrey::sample::interpolate::{Converter, Linear};
 use audrey::sample::signal::{from_iter, Signal};
-use rnnoise_c::DenoiseState;
+use rnnoise_c::{DenoiseState, FRAME_SIZE};
+use structopt::StructOpt;
 
 // RNNoise assumes audio is 16-bit mono with a 48 KHz sample rate
 const SAMPLE_RATE :u32 = 48_000;
 
+#[derive(StructOpt, Debug, Clone)]
+#[structopt(name = "client")]
+pub struct Configuration {
+    /// Verbose mode
+    #[structopt(short, long)]
+    verbose: bool,
+
+    /// Input audio file
+    #[structopt(short, long, required=true)]
+    input: String,
+
+    /// Output audio file
+    #[structopt(short, long, required=true)]
+    output: String,
+}
+
 fn main() {
-    let audio_file_path = args().nth(1)
-        .expect("Please specify an audio file to run RNNoise on");
-    let audio_file = File::open(audio_file_path).unwrap();
+    let configuration = Configuration::from_args();
+    if configuration.verbose {
+        println!("Configuration: {:?}", configuration);
+    }
+
+    let audio_file = File::open(&configuration.input).unwrap();
     let mut reader = Reader::new(audio_file).unwrap();
     let desc = reader.description();
     assert_eq!(1, desc.channel_count(),
         "The channel count is required to be one, at least for now");
 
     // Obtain the buffer of samples
-    let audio_buf :Vec<_> = if desc.sample_rate() == SAMPLE_RATE {
-        reader.samples().map(|s| s.unwrap()).collect()
+    let mut audio_buf :Vec<_> = if desc.sample_rate() == SAMPLE_RATE {
+        reader.samples::<f32>().map(|s| s.unwrap()).collect()
     } else {
         // We need to interpolate to the target sample rate
-        let interpolator = Linear::new([0i16], [0]);
+        let interpolator = Linear::new([0f32], [0.0]);
         let conv = Converter::from_hz_to_hz(
-            from_iter(reader.samples::<i16>().map(|s| [s.unwrap()])),
+            from_iter(reader.samples::<f32>().map(|s| [s.unwrap()])),
             interpolator,
             desc.sample_rate() as f64,
             SAMPLE_RATE as f64);
         conv.until_exhausted().map(|v| v[0]).collect()
     };
 
-    let rnnoise = DenoiseState::new();
+    if configuration.verbose {
+        println!("{} length: {}", &configuration.input, audio_buf.len());
+    }
+
+    // The library requires each frame be exactly FRAME_SIZE, so we append
+    // some zeros to be sure the final frame is sufficiently long.
+    let padding = audio_buf.len() % FRAME_SIZE;
+    if padding > 0 {
+        let mut pad: Vec<f32> = vec![0.0; FRAME_SIZE - padding];
+        audio_buf.append(&mut pad);
+        if configuration.verbose {
+            println!("padded audio file with {} characters", padding);
+        }
+    }
+    let buffers = audio_buf[..].chunks(FRAME_SIZE);
+    let mut denoised_buffer: Vec<f32> = vec![];
+    let mut rnnoise = DenoiseState::new();
+    for buffer in buffers {
+        let mut denoised: Vec<f32> = vec![0.0; FRAME_SIZE];
```

Comment:
Can you create this outside of the loop? It would save allocations.

Reply:
Good point -- I had run into problems earlier and forgot to review closer. It's because I was appending the denoised vector transferring ownership, now I clone it when appending so we only have to allocate once.